### PR TITLE
Use RMM current device resource for memory allocation

### PR DESCRIPTION
**Pull request by tarang-jain**
This PR enables the use of `rmm::mr::current_device_resource` for device allocations when RAFT is enabled. The AllocRequest object caches the rmm Memory Resource used for the allocation to ensure that the same resource is used during deallocation. Managed allocations use a custom default `rmm::mr::managed_memory_resource` because RMM currently does not have a way to retrieve a "guaranteed" managed memory resource from the current device resource.

### CODE REVIEW

**Review by bdice — APPROVED** [review]
This looks fine to me.

**Review by cjnolet — APPROVED** [review]
This LGTM also.

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks @tarang-jain for the PR. There are a few issues, but in general the changes related to the current device resource look good.

File: faiss/gpu/StandardGpuResources.h [hunk 1]
```diff
@@ -166,11 +164,8 @@ class StandardGpuResourcesImpl : public GpuResources {
      * to create a subclass only for the RMM memory resources.
      */
 
-    // cuda_memory_resource
-    std::unique_ptr<rmm::mr::device_memory_resource> cmr;
-
     // managed_memory_resource
-    std::unique_ptr<rmm::mr::device_memory_resource> mmr;
+    rmm::mr::device_memory_resource* mmr;
```
[the remark this answers is not in the document]
Comment (tarang-jain) [reply faiss/gpu/StandardGpuResources.h]: Changed it back to unique pointer. Added the trailing underscore.

File: faiss/gpu/StandardGpuResources.cpp [hunk 2]
```diff
@@ -92,7 +91,6 @@ std::string allocsToString(const std::unordered_map<void*, AllocRequest>& map) {
 StandardGpuResourcesImpl::StandardGpuResourcesImpl()
         :
 #if defined USE_NVIDIA_RAFT
-          cmr(new rmm::mr::cuda_memory_resource),
           mmr(new rmm::mr::managed_memory_resource),
```
[the remark this answers is not in the document]
Comment (tarang-jain) [reply faiss/gpu/StandardGpuResources.cpp]: Changed it back to unique pointer, so the an update to the destructor is not needed.

**Comment by tarang-jain** [comment]
Closing in favour of #3044. These changes have been added to #3044

**Pull request closed without merging**